src/Framework/Constraint/Count.php

```diff
@@ -56,64 +60,106 @@ protected function getCountOf($other): ?int
-                $other = $other->getIterator();
-            }
+            return $this->getCountOfTraversable($other);
+        }
+    }
 
-            $iterator = $other;
+    /**
+     * Returns the description of the failure.
+     *
+     * The beginning of failure messages is "Failed asserting that" in most
+     * cases. This method should return the second part of that sentence.
+     *
+     * @param mixed $other evaluated value or object
+     */
+    protected function failureDescription($other): string
+    {
+        return \sprintf(
+            'actual size %d matches expected size %d',
+            $this->getCountOf($other),
+            $this->expectedCount
+        );
+    }
 
-            if ($iterator instanceof Generator) {
-                return $this->getCountOfGenerator($iterator);
-            }
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->traversableCounts === null) {
+            $this->traversableCounts = new \SplObjectStorage();
+        }
 
-            if (!$iterator instanceof Iterator) {
-                return \iterator_count($iterator);
-            }
+        while ($traversable instanceof IteratorAggregate) {
+            $traversable = $traversable->getIterator();
+        }
 
-            $key   = $iterator->key();
-            $count = \iterator_count($iterator);
+        if ($this->traversableCounts->contains($traversable)) {
+            return $this->traversableCounts[$traversable];
+        }
 
-            // Manually rewind $iterator to previous key, since iterator_count
-            // moves pointer.
-            if ($key !== null) {
-                $iterator->rewind();
+        $count = $this->countNonIteratorAggregateTraversable($traversable);
 
-                while ($iterator->valid() && $key !== $iterator->key()) {
-                    $iterator->next();
-                }
-            }
+        $this->traversableCounts->attach($traversable, $count);
 
-            return $count;
+        return $count;
+    }
+
+    private function countNonIteratorAggregateTraversable(Traversable $traversable): int
+    {
+        if ($traversable instanceof Iterator) {
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
         }
+
+        return \iterator_count($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key          = $iterator->key();
+        $isRewindable = $this->attemptIteratorRewind($iterator);
+        $count        = $this->getCountOfIterator($iterator);
+
+        if ($isRewindable) {
+            $this->moveIteratorToPosition($iterator, $key);
+        }
+
+        return $count;
     }
 
     /**
-     * Returns the total number of iterations from a generator.
-     * This will fully exhaust the generator.
+     * Returns the total number of iterations from a iterator.
```

Comment:
Last comment (got lost when approving, sorry): this should now be "from a**n** iterator".
And maybe you can also improve "the total number of iterations" to the more accurate "the *remaining* number of iterations" or even "the *number of remaining iterations*"?